From 3245a1f5c5911ec522835ae70b1317755e7255e7 Mon Sep 17 00:00:00 2001
From: Igal Liberman <igall@marvell.com>
Date: Sun, 13 Nov 2016 16:45:05 +0200
Subject: [PATCH 0571/1345] phy: comphy: cp110: implement set_mode callback

commit  c412b1a3ff122c0fd4f9ec2496b8ce1f01fc32e4 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

This callback allows the phy consumer to modify the operating mode
of the phy.

Change-Id: I20edbcb1053f91bc979e71ea0c05cf40f2547514
Signed-off-by: Igal Liberman <igall@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/33722
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: Omri Itach <omrii@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/phy/phy-mvebu-comphy.c |   26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/phy/phy-mvebu-comphy.c b/drivers/phy/phy-mvebu-comphy.c
index 8d8375d..a339b70 100644
--- a/drivers/phy/phy-mvebu-comphy.c
+++ b/drivers/phy/phy-mvebu-comphy.c
@@ -671,9 +671,35 @@ static int mvebu_comphy_power_off(struct phy *phy)
 	return 0;
 }
 
+static int mvebu_comphy_set_mode(struct phy *phy, enum phy_mode mode)
+{
+	struct mvebu_comphy *comphy = phy_get_drvdata(phy);
+	struct mvebu_comphy_priv *priv = to_mvebu_comphy_priv(comphy);
+	int i;
+
+	dev_dbg(priv->dev, "%s: Enter\n", __func__);
+
+	for (i = 0; i < MVEBU_COMPHY_FUNC_MAX; i++)
+		if (priv->sinfo->functions[comphy->index][i] == (int)mode)
+			break;
+
+	if (i == MVEBU_COMPHY_FUNC_MAX) {
+		dev_err(priv->dev, "can't set mode 0x%x for COMPHY%d\n",
+			mode, comphy->index);
+		return -EINVAL;
+	}
+
+	priv->lanes[comphy->index].mode = (int)mode;
+
+	dev_dbg(priv->dev, "%s: Exit\n", __func__);
+
+	return 0;
+}
+
 static struct phy_ops mvebu_comphy_ops = {
 	.power_on	= mvebu_comphy_power_on,
 	.power_off	= mvebu_comphy_power_off,
+	.set_mode	= mvebu_comphy_set_mode,
 	.owner		= THIS_MODULE,
 };
 
-- 
1.7.9.5

